### DISCUSSION TITLE
Dependency injection or Inversion of control, Where is this crucial concept in DRF

### DISCUSSION THREAD

**Discussion by kasir-barati** [encode/django-rest-framework#8545] [Potential Issue] [answered]
Hello Community

Came across the dimensions and being in a lovely reliable framework named NestJS I am astonished and also terrified to the core of my fleshes about lacking this important concept in DRF and Django in general. Any suggestion or notion about this topic?

**Comment by lovelydinosaur** [comment, marked as answer]
We aim to astonish. 💖
Hope the core of your flesh recovers nicely.

**Reply by kasir-barati** [reply]
So this means that there is not any intention or plan to add DI in DRF, But why? I guess a good framework like Django and DRF can cover this need too. BTW I am novice in Django and DRF, So maybe I have a grandiose idea about how things should be done. TBH I have a dogmatic believe in NestJS **But** this is not an excuse for me to stop myself from asking why.

This is me, I am eager to learn and ask why. So please bear with me and tell me elaborately why we do not have DI in Django.

I hope I do not make you feel bad. I am just curious.

**Reply by foarsitter** [reply]
An airplane can travel faster than a car, but adding wings to a car doen't make any sense. 

Design patterns are typical solutions to common problems in software design<sup>[1](https://refactoring.guru/design-patterns)</sup>. 

Adding these patterns to an existing codebase without solving any problems does make less sense then adding wings to a car.

**Reply by kasir-barati** [reply]
Interesting way of arguing @foarsitter. But I guess I have to admit it, TBH I cannot say anything as an answer to this quote:
> Design patterns are typical solutions to common problems in software design[1](https://refactoring.guru/design-patterns).

So I guess our discussion is done here and I am mostly convinced that it make no sense to add DI in DRF. But I have to say that I have a great experience in DI while coding in NestJS. And also AFAIK my friend who develop in asp.net also says that it is a whole different world. As you may know .net developers have most of design patterns in place and they just use them.

BTW I am still a fan of DI, especially that it is also useful in unit test and other stuff. mocking and also taking care of Singleton or if you want not Singleton.

In general IDK
